src/WebJobs.Extensions.CosmosDB/Config/DefaultCosmosDBServiceFactory.cs

```diff
@@ -1,15 +1,107 @@
     {
-        public CosmosClient CreateService(string connectionString, CosmosClientOptions cosmosClientOptions)
+        private readonly IConfiguration _configuration;
+        private readonly AzureComponentFactory _componentFactory;
+
+        public DefaultCosmosDBServiceFactory(
+            IConfiguration configuration,
+            AzureComponentFactory componentFactory)
+        {
+            this._configuration = configuration;
+            this._componentFactory = componentFactory;
+        }
+
+        public CosmosClient CreateService(string connectionName, CosmosClientOptions cosmosClientOptions)
+        {
+            CosmosConnectionInformation cosmosConnectionInformation = this.ResolveConnectionInformation(connectionName);
+            if (cosmosConnectionInformation.UsesConnectionString)
+            {
+                // Connection string based auth
+                return new CosmosClient(cosmosConnectionInformation.ConnectionString, cosmosClientOptions);
+            }
+
+            // AAD auth
+            return new CosmosClient(cosmosConnectionInformation.AccountEndpoint, cosmosConnectionInformation.Credential, cosmosClientOptions);
+        }
+
+        private CosmosConnectionInformation ResolveConnectionInformation(string connection)
         {
-            return new CosmosClient(connectionString, cosmosClientOptions);
+            var connectionSetting = connection ?? Constants.DefaultConnectionStringName;
+            IConfigurationSection connectionSection = GetWebJobsConnectionStringSectionCosmos(this._configuration, connectionSetting);
+            if (!connectionSection.Exists())
+            {
+                // Not found
+                throw new InvalidOperationException($"Cosmos DB connection configuration '{connectionSetting}' does not exist. " +
+                                                    $"Make sure that it is a defined App Setting.");
+            }
+
+            if (!string.IsNullOrWhiteSpace(connectionSection.Value))
+            {
+                return new CosmosConnectionInformation(connectionSection.Value);
+            }
+            else
+            {
+                string accountEndpoint = connectionSection["accountEndpoint"];
+                if (string.IsNullOrWhiteSpace(accountEndpoint))
+                {
+                    // Not found
+                    throw new InvalidOperationException($"Connection should have an 'accountEndpoint' property or be a " +
+                        $"string representing a connection string.");
+                }
+
+                TokenCredential credential = _componentFactory.CreateTokenCredential(connectionSection);
+                return new CosmosConnectionInformation(accountEndpoint, credential);
+            }
+        }
+
+        public static IConfigurationSection GetWebJobsConnectionStringSectionCosmos(IConfiguration configuration, string connectionStringName)
```

Comment:
Why not use the GetWebJobsConnectionStringSection from `WebJobsConfigurationExtensions.cs`

Reply:
@brettsam had the same question, the problem is, `WebJobsConfigurationExtensions` resolve using the WebJobs root path `AzureWebJobs:extensions:<config name>`: https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Extensions/IConfigurationExtensions.cs#L42-L46

The other method available `GetConnectionStringOrSetting` only returns a string, not the Section, so we cannot use it.

If we take a look at the other repository uses these methods, which are the ones I copied: https://github.com/Azure/azure-sdk-for-net/blob/303b75341ba17aecf239a3232d28ca76209a5346/sdk/extensions/Microsoft.Azure.WebJobs.Extensions.Clients/src/Shared/WebJobsConfigurationExtensions.cs#L12